File: src/services/mythrad/wire/wire.go
// Package wire is the Wake Up Mythra session protocol: length-prefixed
// binary frames on the one bidi stream, fixed-layout datagrams beside it.
// Every field is little-endian except the frame length, a QUIC
// variable-length integer (RFC 9000 §16), which is big-endian by that
// spec's definition. Three implementations — this one, the Rust session
// core, and the TS host — must produce byte-identical output, so
// wire_test.go pins golden vectors rather than only round-trips.
package wire

import (
	"bufio"
	"encoding/binary"
	"encoding/hex"
	"errors"
	"io"
)

const Proto = 4

// Frame kinds. 1-15 travel client to server, 16 and up server to client.
const (
	KindHello  = 1
	KindIntent = 2
	KindResync = 3

	KindWelcome  = 16
	KindEvent    = 17
	KindReject   = 18
	KindSnapshot = 19
)

// Datagram kinds. A datagram's boundary is its message boundary, so these
// carry no length prefix — only a leading kind byte and a fixed layout.
const (
	DgCheck   = 1
	DgVerdict = 2

	CheckLen   = 25
	VerdictLen = 34
)

// MaxFrameLen bounds one frame body (kind byte plus payload). A deflated
// park snapshot sits far below it; anything larger is a peer that has lost
// the framing, not a big message.
const MaxFrameLen = 128 * 1024

// Verdict flags: Known says the tick was still in the authority's hash
// ring, OK answers the comparison and is meaningful only alongside Known.
const (
	VerdictKnown = 1 << 0
	VerdictOK    = 1 << 1
)

const (
	RoleSpectator = 0
	RolePlayer    = 1
)

var (
	ErrShortFrame    = errors.New("frame carries no kind byte")
	ErrFrameTooLarge = errors.New("frame exceeds the size cap")
	ErrBadPayload    = errors.New("frame payload does not match its layout")
	ErrBadDatagram   = errors.New("datagram does not match its layout")
)

// ---------- variable-length integers ----------

func VarintLen(v uint64) int {
	switch {
	case v <= 63:
		return 1
	case v <= 16383:
		return 2
	case v <= 1073741823:
		return 4
	default:
		return 8
	}
}

// AppendVarint writes v in the shortest of RFC 9000 §16's four forms: a
// two-bit length prefix followed by the value in network byte order.
// Encoding is always shortest-form so frames are byte-stable across
// implementations; decoding accepts any form.
func AppendVarint(b []byte, v uint64) []byte {
	switch VarintLen(v) {
	case 1:
		return append(b, byte(v))
	case 2:
		return append(b, byte(v>>8)|0x40, byte(v))
	case 4:
		return append(b, byte(v>>24)|0x80, byte(v>>16), byte(v>>8), byte(v))
	default:
		return append(b, byte(v>>56)|0xc0, byte(v>>48), byte(v>>40), byte(v>>32),
			byte(v>>24), byte(v>>16), byte(v>>8), byte(v))
	}
}

func ReadVarint(r *bufio.Reader) (uint64, error) {
	first, err := r.ReadByte()
	if err != nil {
		return 0, err
	}
	v := uint64(first & 0x3f)
	for i := 1; i < 1<<(first>>6); i++ {
		b, err := r.ReadByte()
		if err != nil {
			return 0, err
		}
		v = v<<8 | uint64(b)
	}
	return v, nil
}

// FrameSize is the on-wire byte count of a frame carrying payloadLen bytes
// of payload, for callers accounting for bandwidth.
func FrameSize(payloadLen int) int {
	body := payloadLen + 1
	return VarintLen(uint64(body)) + body
}

// ---------- framing ----------

// startFrame lays down the length prefix and kind for a body of exactly
// payloadLen+1 bytes, sizing the buffer so the whole frame is one
// allocation.
func startFrame(kind byte, payloadLen int) []byte {
	body := payloadLen + 1
	b := make([]byte, 0, VarintLen(uint64(body))+body)
	return append(AppendVarint(b, uint64(body)), kind)
}

// Reader decodes the frame stream. Payloads are freshly allocated:
// intents outlive the read, staged for a tick boundary and journaled.
type Reader struct {
	r *bufio.Reader
}

func NewReader(r io.Reader) *Reader {
	return &Reader{r: bufio.NewReaderSize(r, 8*1024)}
}

// Next returns the next frame's kind and payload.
func (fr *Reader) Next() (byte, []byte, error) {
	n, err := ReadVarint(fr.r)
	if err != nil {
		return 0, nil, err
	}
	switch {
	case n == 0:
		return 0, nil, ErrShortFrame
	case n > MaxFrameLen:
		return 0, nil, ErrFrameTooLarge
	}
	body := make([]byte, n)
	if _, err := io.ReadFull(fr.r, body); err != nil {
		return 0, nil, err
	}
	return body[0], body[1:], nil
}

// ---------- payload cursor ----------

var zeroField [8]byte

// cursor reads fixed-width fields, latching overrun so a decoder can lay
// its fields out in order and bounds-check once at the end.
type cursor struct {
	b   []byte
	bad bool
}

func (c *cursor) take(n int) []byte {
	if c.bad || len(c.b) < n {
		c.bad = true
		return zeroField[:n]
	}
	out := c.b[:n]
	c.b = c.b[n:]
	return out
}

func (c *cursor) u8() uint8   { return c.take(1)[0] }
func (c *cursor) u16() uint16 { return binary.LittleEndian.Uint16(c.take(2)) }
func (c *cursor) u32() uint32 { return binary.LittleEndian.Uint32(c.take(4)) }
func (c *cursor) u64() uint64 { return binary.LittleEndian.Uint64(c.take(8)) }
func (c *cursor) i64() int64  { return int64(c.u64()) }

func (c *cursor) bytes(n int) []byte {
	if c.bad || len(c.b) < n {
		c.bad = true
		return nil
	}
	out := c.b[:n:n]
	c.b = c.b[n:]
	return out
}

// done reports a clean decode: every field present and nothing trailing.
// Trailing bytes are an encoder disagreeing with this layout, which is
// exactly the failure the three implementations must not paper over.
func (c *cursor) done() bool { return !c.bad && len(c.b) == 0 }

// ---------- client to server ----------

type Hello struct {
	Proto     uint16
	SinceSeq  int64
	SinceTick uint64
	Ticket    string
}

func EncodeHello(f Hello) []byte {
	b := startFrame(KindHello, 20+len(f.Ticket))
	b = binary.LittleEndian.AppendUint16(b, f.Proto)
	b = binary.LittleEndian.AppendUint64(b, uint64(f.SinceSeq))
	b = binary.LittleEndian.AppendUint64(b, f.SinceTick)
	b = binary.LittleEndian.AppendUint16(b, uint16(len(f.Ticket)))
	return append(b, f.Ticket...)
}

func DecodeHello(p []byte) (Hello, error) {
	c := cursor{b: p}
	var f Hello
	f.Proto = c.u16()
	f.SinceSeq = c.i64()
	f.SinceTick = c.u64()
	f.Ticket = string(c.bytes(int(c.u16())))
	if !c.done() {
		return Hello{}, ErrBadPayload
	}
	return f, nil
}

type Intent struct {
	ID      uint64
	Kind    uint16
	Payload []byte
}

func EncodeIntent(f Intent) []byte {
	b := startFrame(KindIntent, 12+len(f.Payload))
	b = binary.LittleEndian.AppendUint64(b, f.ID)
	b = binary.LittleEndian.AppendUint16(b, f.Kind)
	b = binary.LittleEndian.AppendUint16(b, uint16(len(f.Payload)))
	return append(b, f.Payload...)
}

func DecodeIntent(p []byte) (Intent, error) {
	c := cursor{b: p}
	var f Intent
	f.ID = c.u64()
	f.Kind = c.u16()
	f.Payload = c.bytes(int(c.u16()))
	if !c.done() {
		return Intent{}, ErrBadPayload
	}
	return f, nil
}

type Resync struct {
	HaveSeq int64
}

func EncodeResync(f Resync) []byte {
	return binary.LittleEndian.AppendUint64(startFrame(KindResync, 8), uint64(f.HaveSeq))
}

func DecodeResync(p []byte) (Resync, error) {
	c := cursor{b: p}
	f := Resync{HaveSeq: c.i64()}
	if !c.done() {
		return Resync{}, ErrBadPayload
	}
	return f, nil
}

// ---------- server to client ----------

// Welcome opens every session. Terrain is the blob id itself; hosts render
// it as hex for /terrain/<id>. Hz is a read-out of the world's journaled
// rate, which only ever changes while the park is dark, so a connection
// sees exactly one.
type Welcome struct {
	Epoch   uint32
	Seq     int64
	Tick    uint64
	Hz      uint32
	Role    uint8
	Terrain uint64
	Park    string
}

func EncodeWelcome(f Welcome) []byte {
	park := f.Park
	if len(park) > 255 {
		park = park[:255]
	}
	b := startFrame(KindWelcome, 34+len(park))
	b = binary.LittleEndian.AppendUint32(b, f.Epoch)
	b = binary.LittleEndian.AppendUint64(b, uint64(f.Seq))
	b = binary.LittleEndian.AppendUint64(b, f.Tick)
	b = binary.LittleEndian.AppendUint32(b, f.Hz)
	b = append(b, f.Role)
	b = binary.LittleEndian.AppendUint64(b, f.Terrain)
	b = append(b, byte(len(park)))
	return append(b, park...)
}

func DecodeWelcome(p []byte) (Welcome, error) {
	c := cursor{b: p}
	var f Welcome
	f.Epoch = c.u32()
	f.Seq = c.i64()
	f.Tick = c.u64()
	f.Hz = c.u32()
	f.Role = c.u8()
	f.Terrain = c.u64()
	f.Park = string(c.bytes(int(c.u8())))
	if !c.done() {
		return Welcome{}, ErrBadPayload
	}
	return f, nil
}

// Event is a journal event as peers see it. The actor stays in the
// journal: authorship is a durable fact, not something every replica needs
// to apply the event. Consumers matching their own submissions do it by
// Intent, which must therefore be unique per sender.
type Event struct {
	Seq     int64
	Tick    uint64
	Kind    uint16
	Intent  uint64
	Payload []byte
}

func EncodeEvent(f Event) []byte {
	b := startFrame(KindEvent, 28+len(f.Payload))
	b = binary.LittleEndian.AppendUint64(b, uint64(f.Seq))
	b = binary.LittleEndian.AppendUint64(b, f.Tick)
	b = binary.LittleEndian.AppendUint16(b, f.Kind)
	b = binary.LittleEndian.AppendUint64(b, f.Intent)
	b = binary.LittleEndian.AppendUint16(b, uint16(len(f.Payload)))
	return append(b, f.Payload...)
}

func DecodeEvent(p []byte) (Event, error) {
	c := cursor{b: p}
	var f Event
	f.Seq = c.i64()
	f.Tick = c.u64()
	f.Kind = c.u16()
	f.Intent = c.u64()
	f.Payload = c.bytes(int(c.u16()))
	if !c.done() {
		return Event{}, ErrBadPayload
	}
	return f, nil
}

type Reject struct {
	Intent uint64
	Reason uint32
}

func EncodeReject(f Reject) []byte {
	b := startFrame(KindReject, 12)
	b = binary.LittleEndian.AppendUint64(b, f.Intent)
	return binary.LittleEndian.AppendUint32(b, f.Reason)
}

func DecodeReject(p []byte) (Reject, error) {
	c := cursor{b: p}
	var f Reject
	f.Intent = c.u64()
	f.Reason = c.u32()
	if !c.done() {
		return Reject{}, ErrBadPayload
	}
	return f, nil
}

// Snapshot carries raw deflate of the module's state blob.
type Snapshot struct {
	Seq     int64
	Tick    uint64
	Epoch   uint32
	WH      uint64
	Terrain uint64
	Z       []byte
}

func EncodeSnapshot(f Snapshot) []byte {
	b := startFrame(KindSnapshot, 40+len(f.Z))
	b = binary.LittleEndian.AppendUint64(b, uint64(f.Seq))
	b = binary.LittleEndian.AppendUint64(b, f.Tick)
	b = binary.LittleEndian.AppendUint32(b, f.Epoch)
	b = binary.LittleEndian.AppendUint64(b, f.WH)
	b = binary.LittleEndian.AppendUint64(b, f.Terrain)
	b = binary.LittleEndian.AppendUint32(b, uint32(len(f.Z)))
	return append(b, f.Z...)
}

func DecodeSnapshot(p []byte) (Snapshot, error) {
	c := cursor{b: p}
	var f Snapshot
	f.Seq = c.i64()
	f.Tick = c.u64()
	f.Epoch = c.u32()
	f.WH = c.u64()
	f.Terrain = c.u64()
	n := c.u32()
	if c.bad || uint64(n) > uint64(len(c.b)) {
		return Snapshot{}, ErrBadPayload
	}
	f.Z = c.bytes(int(n))
	if !c.done() {
		return Snapshot{}, ErrBadPayload
	}
	return f, nil
}

// ---------- datagrams ----------

type Check struct {
	Tick uint64
	WH   uint64
	CTMS uint64
}

func EncodeCheck(d Check) []byte {
	b := make([]byte, 0, CheckLen)
	b = append(b, DgCheck)
	b = binary.LittleEndian.AppendUint64(b, d.Tick)
	b = binary.LittleEndian.AppendUint64(b, d.WH)
	return binary.LittleEndian.AppendUint64(b, d.CTMS)
}

func DecodeCheck(b []byte) (Check, error) {
	if len(b) != CheckLen || b[0] != DgCheck {
		return Check{}, ErrBadDatagram
	}
	c := cursor{b: b[1:]}
	return Check{Tick: c.u64(), WH: c.u64(), CTMS: c.u64()}, nil
}

// Verdict answers a check and rides the current module hashes, so a page
// learns mid-session that its wasm is stale. CW/PW hold the first four
// bytes of each module's sha256 — the bytes the 8-hex-char display form
// spells out, in that order.
type Verdict struct {
	Tick  uint64
	Now   uint64
	CTMS  uint64
	Flags uint8
	CW    uint32
	PW    uint32
}

func EncodeVerdict(d Verdict) []byte {
	b := make([]byte, 0, VerdictLen)
	b = append(b, DgVerdict)
	b = binary.LittleEndian.AppendUint64(b, d.Tick)
	b = binary.LittleEndian.AppendUint64(b, d.Now)
	b = binary.LittleEndian.AppendUint64(b, d.CTMS)
	b = append(b, d.Flags)
	b = binary.LittleEndian.AppendUint32(b, d.CW)
	return binary.LittleEndian.AppendUint32(b, d.PW)
}

func DecodeVerdict(b []byte) (Verdict, error) {
	if len(b) != VerdictLen || b[0] != DgVerdict {
		return Verdict{}, ErrBadDatagram
	}
	c := cursor{b: b[1:]}
	return Verdict{
		Tick: c.u64(), Now: c.u64(), CTMS: c.u64(),
		Flags: c.u8(), CW: c.u32(), PW: c.u32(),
	}, nil
}

// ModuleWord packs a distribution hash for a verdict. The display form is
// hex of the sha256's first four bytes, so decoding it and reading those
// bytes little-endian puts them back on the wire in their original order.
func ModuleWord(hash string) uint32 {
	var b [4]byte
	if _, err := hex.Decode(b[:], []byte(hash)); err != nil {
		return 0
	}
	return binary.LittleEndian.Uint32(b[:])
}

// HexWord is ModuleWord's inverse: the 8-character display form.
func HexWord(word uint32) string {
	var b [4]byte
	binary.LittleEndian.PutUint32(b[:], word)
	return hex.EncodeToString(b[:])
}

func RoleCode(role string) uint8 {
	if role == "player" {
		return RolePlayer
	}
	return RoleSpectator
}

File: src/services/mythrad/sim/client/session/src/wire.rs
//! Wire protocol v4: length-prefixed binary frames on one bidirectional
//! stream, plus unframed check/verdict datagrams. Both directions encode
//! and decode here, so one definition answers for the replica, the
//! harnesses, and the conformance vectors the Go authority is held to.
//!
//! Every scalar is little-endian. The one exception is the frame's QUIC
//! variable-length integer length prefix, which is big-endian because
//! RFC 9000 §16 says so.

/// Client to server.
pub const K_HELLO: u8 = 1;
pub const K_INTENT: u8 = 2;
pub const K_RESYNC: u8 = 3;
/// Server to client.
pub const K_WELCOME: u8 = 16;
pub const K_EVENT: u8 = 17;
pub const K_REJECT: u8 = 18;
pub const K_SNAPSHOT: u8 = 19;

pub const DG_CHECK: u8 = 1;
pub const DG_VERDICT: u8 = 2;

pub const PROTO: u16 = 4;

pub const HELLO_HEADER: usize = 20;
pub const INTENT_HEADER: usize = 12;
pub const RESYNC_BYTES: usize = 8;
pub const WELCOME_HEADER: usize = 34;
pub const EVENT_HEADER: usize = 28;
pub const REJECT_BYTES: usize = 12;
pub const SNAPSHOT_HEADER: usize = 40;
pub const CHECK_BYTES: usize = 25;
pub const VERDICT_BYTES: usize = 34;

/// The checked tick was still inside the authority's hash ring.
pub const VERDICT_KNOWN: u8 = 1;
/// The hashes agreed — meaningful only alongside `VERDICT_KNOWN`.
pub const VERDICT_OK: u8 = 1 << 1;

pub const VARINT_MAX: u64 = (1 << 62) - 1;

/// The u32 the session ABI carries for a module word: the little-endian
/// load of the four wire bytes, so `module_word(w).to_le_bytes() == w`.
pub fn module_word(bytes: [u8; 4]) -> u32 {
    u32::from_le_bytes(bytes)
}

pub fn varint_len(v: u64) -> usize {
    match v {
        0..=0x3F => 1,
        0x40..=0x3FFF => 2,
        0x4000..=0x3FFF_FFFF => 4,
        _ => 8,
    }
}

/// Writes the shortest form of `v`, returning its byte length.
pub fn put_varint(dst: &mut [u8], v: u64) -> usize {
    match varint_len(v) {
        1 => {
            dst[0] = v as u8;
            1
        }
        2 => {
            dst[..2].copy_from_slice(&((v as u16) | 0x4000).to_be_bytes());
            2
        }
        4 => {
            dst[..4].copy_from_slice(&((v as u32) | 0x8000_0000).to_be_bytes());
            4
        }
        _ => {
            dst[..8].copy_from_slice(&(v | (3 << 62)).to_be_bytes());
            8
        }
    }
}

/// Reads a varint, or `None` when fewer than its encoded length remain —
/// longer-than-canonical forms decode fine, as RFC 9000 permits.
pub fn get_varint(src: &[u8]) -> Option<(u64, usize)> {
    let first = *src.first()?;
    let len = 1usize << (first >> 6);
    if src.len() < len {
        return None;
    }
    let mut b = [0u8; 8];
    b[8 - len..].copy_from_slice(&src[..len]);
    let v = u64::from_be_bytes(b) & (!0u64 >> (64 - 6 - 8 * (len as u32 - 1)));
    Some((v, len))
}

fn u16le(b: &[u8], at: usize) -> u16 {
    u16::from_le_bytes([b[at], b[at + 1]])
}

fn u32le(b: &[u8], at: usize) -> u32 {
    u32::from_le_bytes([b[at], b[at + 1], b[at + 2], b[at + 3]])
}

fn u64le(b: &[u8], at: usize) -> u64 {
    let mut v = [0u8; 8];
    v.copy_from_slice(&b[at..at + 8]);
    u64::from_le_bytes(v)
}

/// Locates the first whole frame in `buf`: `(kind, payload offset, payload
/// length, total frame length)`. `Ok(None)` means "need more bytes";
/// `Err(())` means the stream is unreadable and the connection is done.
#[allow(clippy::type_complexity)]
pub fn frame_bounds(buf: &[u8]) -> Result<Option<(u8, usize, usize, usize)>, ()> {
    let Some((qlen, pre)) = get_varint(buf) else {
        return Ok(None);
    };
    if qlen == 0 || qlen > VARINT_MAX {
        return Err(());
    }
    let total = pre + qlen as usize;
    if buf.len() < total {
        return Ok(None);
    }
    Ok(Some((buf[pre], pre + 1, qlen as usize - 1, total)))
}

/// Writes the length prefix and kind, returning `(payload offset, total)`.
fn frame(dst: &mut [u8], kind: u8, payload_len: usize) -> (usize, usize) {
    let body = 1 + payload_len;
    let pre = put_varint(dst, body as u64);
    dst[pre] = kind;
    (pre + 1, pre + body)
}

pub fn encode_hello(dst: &mut [u8], since_seq: i64, since_tick: u64, ticket: &[u8]) -> usize {
    let (at, total) = frame(dst, K_HELLO, HELLO_HEADER + ticket.len());
    dst[at..at + 2].copy_from_slice(&PROTO.to_le_bytes());
    dst[at + 2..at + 10].copy_from_slice(&since_seq.to_le_bytes());
    dst[at + 10..at + 18].copy_from_slice(&since_tick.to_le_bytes());
    dst[at + 18..at + 20].copy_from_slice(&(ticket.len() as u16).to_le_bytes());
    dst[at + 20..at + 20 + ticket.len()].copy_from_slice(ticket);
    total
}

pub fn encode_intent(dst: &mut [u8], id: u64, kind: u16, p: &[u8]) -> usize {
    let (at, total) = frame(dst, K_INTENT, INTENT_HEADER + p.len());
    dst[at..at + 8].copy_from_slice(&id.to_le_bytes());
    dst[at + 8..at + 10].copy_from_slice(&kind.to_le_bytes());
    dst[at + 10..at + 12].copy_from_slice(&(p.len() as u16).to_le_bytes());
    dst[at + 12..at + 12 + p.len()].copy_from_slice(p);
    total
}

pub fn encode_resync(dst: &mut [u8], have_seq: i64) -> usize {
    let (at, total) = frame(dst, K_RESYNC, RESYNC_BYTES);
    dst[at..at + 8].copy_from_slice(&have_seq.to_le_bytes());
    total
}

pub fn encode_welcome(dst: &mut [u8], w: &Welcome) -> usize {
    let (at, total) = frame(dst, K_WELCOME, WELCOME_HEADER + w.park.len());
    dst[at..at + 4].copy_from_slice(&w.epoch.to_le_bytes());
    dst[at + 4..at + 12].copy_from_slice(&w.seq.to_le_bytes());
    dst[at + 12..at + 20].copy_from_slice(&w.tick.to_le_bytes());
    dst[at + 20..at + 24].copy_from_slice(&w.hz.to_le_bytes());
    dst[at + 24] = w.role;
    dst[at + 25..at + 33].copy_from_slice(&w.terrain.to_le_bytes());
    dst[at + 33] = w.park.len() as u8;
    dst[at + 34..at + 34 + w.park.len()].copy_from_slice(w.park);
    total
}

pub fn encode_event(dst: &mut [u8], e: &Event) -> usize {
    let (at, total) = frame(dst, K_EVENT, EVENT_HEADER + e.p.len());
    dst[at..at + 8].copy_from_slice(&e.seq.to_le_bytes());
    dst[at + 8..at + 16].copy_from_slice(&e.tick.to_le_bytes());
    dst[at + 16..at + 18].copy_from_slice(&e.kind.to_le_bytes());
    dst[at + 18..at + 26].copy_from_slice(&e.intent.to_le_bytes());
    dst[at + 26..at + 28].copy_from_slice(&(e.p.len() as u16).to_le_bytes());
    dst[at + 28..at + 28 + e.p.len()].copy_from_slice(e.p);
    total
}

pub fn encode_reject(dst: &mut [u8], intent: u64, reason: u32) -> usize {
    let (at, total) = frame(dst, K_REJECT, REJECT_BYTES);
    dst[at..at + 8].copy_from_slice(&intent.to_le_bytes());
    dst[at + 8..at + 12].copy_from_slice(&reason.to_le_bytes());
    total
}

pub fn encode_snapshot(dst: &mut [u8], s: &Snapshot) -> usize {
    let (at, total) = frame(dst, K_SNAPSHOT, SNAPSHOT_HEADER + s.z.len());
    dst[at..at + 8].copy_from_slice(&s.seq.to_le_bytes());
    dst[at + 8..at + 16].copy_from_slice(&s.tick.to_le_bytes());
    dst[at + 16..at + 20].copy_from_slice(&s.epoch.to_le_bytes());
    dst[at + 20..at + 28].copy_from_slice(&s.wh.to_le_bytes());
    dst[at + 28..at + 36].copy_from_slice(&s.terrain.to_le_bytes());
    dst[at + 36..at + 40].copy_from_slice(&(s.z.len() as u32).to_le_bytes());
    dst[at + 40..at + 40 + s.z.len()].copy_from_slice(s.z);
    total
}

pub fn encode_check(dst: &mut [u8], tick: u64, wh: u64, ct_ms: u64) -> usize {
    dst[0] = DG_CHECK;
    dst[1..9].copy_from_slice(&tick.to_le_bytes());
    dst[9..17].copy_from_slice(&wh.to_le_bytes());
    dst[17..25].copy_from_slice(&ct_ms.to_le_bytes());
    CHECK_BYTES
}

pub fn encode_verdict(dst: &mut [u8], v: &Verdict) -> usize {
    dst[0] = DG_VERDICT;
    dst[1..9].copy_from_slice(&v.tick.to_le_bytes());
    dst[9..17].copy_from_slice(&v.now.to_le_bytes());
    dst[17..25].copy_from_slice(&v.ct_ms.to_le_bytes());
    dst[25] = v.flags;
    dst[26..30].copy_from_slice(&v.cw);
    dst[30..34].copy_from_slice(&v.pw);
    VERDICT_BYTES
}

pub struct Hello<'a> {
    pub proto: u16,
    pub since_seq: i64,
    pub since_tick: u64,
    pub ticket: &'a [u8],
}

pub struct Intent<'a> {
    pub id: u64,
    pub kind: u16,
    pub p: &'a [u8],
}

pub struct Welcome<'a> {
    pub epoch: u32,
    pub seq: i64,
    pub tick: u64,
    pub hz: u32,
    pub role: u8,
    pub terrain: u64,
    pub park: &'a [u8],
}

pub struct Event<'a> {
    pub seq: i64,
    pub tick: u64,
    pub kind: u16,
    pub intent: u64,
    pub p: &'a [u8],
}

pub struct Reject {
    pub intent: u64,
    pub reason: u32,
}

pub struct Snapshot<'a> {
    pub seq: i64,
    pub tick: u64,
    pub epoch: u32,
    pub wh: u64,
    pub terrain: u64,
    pub z: &'a [u8],
}

pub struct Check {
    pub tick: u64,
    pub wh: u64,
    pub ct_ms: u64,
}

pub struct Verdict {
    pub tick: u64,
    pub now: u64,
    pub ct_ms: u64,
    pub flags: u8,
    /// The client and park module sha256 prefixes, verbatim and in display
    /// order: hexing these left to right reproduces what `/wt-info` shows.
    /// Only the session ABI turns one into a number, and only ever as the
    /// little-endian load of these same four bytes.
    pub cw: [u8; 4],
    pub pw: [u8; 4],
}

pub fn parse_hello(p: &[u8]) -> Option<Hello<'_>> {
    if p.len() < HELLO_HEADER {
        return None;
    }
    let tlen = u16le(p, 18) as usize;
    if p.len() != HELLO_HEADER + tlen {
        return None;
    }
    Some(Hello {
        proto: u16le(p, 0),
        since_seq: u64le(p, 2) as i64,
        since_tick: u64le(p, 10),
        ticket: &p[20..20 + tlen],
    })
}

pub fn parse_intent(p: &[u8]) -> Option<Intent<'_>> {
    if p.len() < INTENT_HEADER {
        return None;
    }
    let plen = u16le(p, 10) as usize;
    if p.len() != INTENT_HEADER + plen {
        return None;
    }
    Some(Intent {
        id: u64le(p, 0),
        kind: u16le(p, 8),
        p: &p[12..12 + plen],
    })
}

pub fn parse_resync(p: &[u8]) -> Option<i64> {
    (p.len() == RESYNC_BYTES).then(|| u64le(p, 0) as i64)
}

pub fn parse_welcome(p: &[u8]) -> Option<Welcome<'_>> {
    if p.len() < WELCOME_HEADER {
        return None;
    }
    let nlen = p[33] as usize;
    if p.len() != WELCOME_HEADER + nlen {
        return None;
    }
    Some(Welcome {
        epoch: u32le(p, 0),
        seq: u64le(p, 4) as i64,
        tick: u64le(p, 12),
        hz: u32le(p, 20),
        role: p[24],
        terrain: u64le(p, 25),
        park: &p[34..34 + nlen],
    })
}

pub fn parse_event(p: &[u8]) -> Option<Event<'_>> {
    if p.len() < EVENT_HEADER {
        return None;
    }
    let plen = u16le(p, 26) as usize;
    if p.len() != EVENT_HEADER + plen {
        return None;
    }
    Some(Event {
        seq: u64le(p, 0) as i64,
        tick: u64le(p, 8),
        kind: u16le(p, 16),
        intent: u64le(p, 18),
        p: &p[28..28 + plen],
    })
}

pub fn parse_reject(p: &[u8]) -> Option<Reject> {
    (p.len() == REJECT_BYTES).then(|| Reject {
        intent: u64le(p, 0),
        reason: u32le(p, 8),
    })
}

pub fn parse_snapshot(p: &[u8]) -> Option<Snapshot<'_>> {
    if p.len() < SNAPSHOT_HEADER {
        return None;
    }
    let zlen = u32le(p, 36) as usize;
    if p.len() != SNAPSHOT_HEADER + zlen {
        return None;
    }
    Some(Snapshot {
        seq: u64le(p, 0) as i64,
        tick: u64le(p, 8),
        epoch: u32le(p, 16),
        wh: u64le(p, 20),
        terrain: u64le(p, 28),
        z: &p[40..40 + zlen],
    })
}

pub fn parse_check(b: &[u8]) -> Option<Check> {
    if b.len() != CHECK_BYTES || b[0] != DG_CHECK {
        return None;
    }
    Some(Check {
        tick: u64le(b, 1),
        wh: u64le(b, 9),
        ct_ms: u64le(b, 17),
    })
}

pub fn parse_verdict(b: &[u8]) -> Option<Verdict> {
    if b.len() != VERDICT_BYTES || b[0] != DG_VERDICT {
        return None;
    }
    Some(Verdict {
        tick: u64le(b, 1),
        now: u64le(b, 9),
        ct_ms: u64le(b, 17),
        flags: b[25],
        cw: [b[26], b[27], b[28], b[29]],
        pw: [b[30], b[31], b[32], b[33]],
    })
}
